### PR TITLE
Architecture: Implement Heap

### PR DESCRIPTION
**Pull request by talekhinezh**
* Implement `Heap` store, flattened out store to contain flat set of nodes
* Move node movement logic into `CallFrame` + `Heap`
* Move ownership and reference handling logic into `CallFrame`
* Remove floating `RENodePointer`, all substate access is done via `CallFrame` now
* Cleanup AuthModule initialization

### CODE REVIEW

File: radix-engine/src/engine/call_frame.rs [hunk 1]
```diff
@@ -20,50 +28,86 @@ pub struct CallFrame {
...
-    /// 1. borrowed values
-    /// 2. track
-    pub node_refs: HashMap<RENodeId, RENodePointer>,
+    /// All ref nodes accessible by this call frame (does not include owned nodes).
+    pub node_refs: HashMap<RENodeId, RENodeLocation>,
 
-    /// Owned Values
-    pub owned_heap_nodes: HashMap<RENodeId, HeapRootRENode>,
+    /// Owned nodes which by definition must live on heap
+    owned_heap_nodes: HashSet<RENodeId>,
```
Comment (iamyulong) [inline radix-engine/src/engine/call_frame.rs:35]: Suggest renaming to `owned_root_nodes`

File: radix-engine/src/engine/call_frame.rs [hunk 2]
```diff
@@ -21,50 +29,86 @@ pub struct CallFrame {
...
         offset: SubstateOffset,
         flags: LockFlags,
-    ) -> LockHandle {
+    ) -> Result<LockHandle, RuntimeError> {
+        let location = self.get_node_location(node_id)?;
+        if !(matches!(offset, SubstateOffset::KeyValueStore(..))
+            || matches!(
+                offset,
+                SubstateOffset::NonFungibleStore(NonFungibleStoreOffset::Entry(..))
+            ))
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/call_frame.rs:55]: They will be handled the same once work on virtual key/value store is done so I don't think we need to change this

File: radix-engine/src/engine/call_frame.rs [hunk 3]
```diff
@@ -118,94 +178,350 @@ impl CallFrame {
...
+        children: HashSet<RENodeId>,
+        to: RENodeId,
+    ) -> Result<(), RuntimeError> {
+        for child_id in &children {
+            self.take_node_internal(heap, *child_id)?;
+        }
+        heap.add_child_nodes(children, to)?;
+
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/call_frame.rs:350]: good point

File: radix-engine/src/engine/heap.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    pub fn add_child_nodes(
+        &mut self,
+        node_ids: HashSet<RENodeId>,
+        to: RENodeId,
+    ) -> Result<(), CallFrameError> {
+        for node_id in &node_ids {
+            // Sanity check
+            if !self.nodes.contains_key(&node_id) {
+                return Err(CallFrameError::RENodeNotOwned(*node_id));
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/heap.rs:65]: This is just a sanity check. The `CallFrame` is the one who should be managing correctness. I'm thinking that we create a new error like `InvariantBrokenError` which would catch broken invariants like this. What do you think?

File: radix-engine/src/engine/heap.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        self.move_nodes_to_store(track, node.child_nodes)?;
+
+        Ok(())
+    }
+
+    pub fn remove_node(&mut self, node_id: RENodeId) -> Result<HeapRENode, CallFrameError> {
+        self.nodes
+            .remove(&node_id)
+            .ok_or(CallFrameError::RENodeNotOwned(node_id))
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/heap.rs:112]: This is also an invariant broken error

File: radix-engine/src/engine/heap.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/engine/heap.rs:65]: I think a better approach is to move the parent-child relationship management to the client, if the heap is not going to check that.

In this case, the `Heap` returns a mut ref to the `node/child_ids`, so clients can do whatever they like.

[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/heap.rs:65]: a good point!

[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/heap.rs:65]: Can we defer that change to a later PR though

**Review by iamyulong — APPROVED** [review]

File: radix-engine/src/engine/heap.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/heap.rs:65]: Actually I don't think this works as we may need to move nodes owned by one call frame to another node owned by another call frame.

**Pull request merged**